File: apps/swift-ui-app/Sources/LeoGit/Services/RepoSearch.swift
import Foundation

/// Ranked fuzzy search over the repository list.
///
/// The Tauri picker's rule — the query is a subsequence of the name, the
/// `owner/name` label, *or the full path* — collapses under a real scan tree.
/// Every row shares the same ancestry, so a short query finds its letters
/// scattered through that shared prefix and matches everything: under
/// `/Users/leo/Dev/LeoManrique/Desktop`, `llm` matched all fourteen
/// repositories, reading as the `l` of `leo`, the `l` of `LeoManrique`, and
/// that word's `m`.
///
/// So the two halves of a path are searched by different rules. The name keeps
/// the scattered-subsequence match a fuzzy finder is expected to have, while
/// the path must contain the query *contiguously* and is first trimmed to what
/// lies below the scan folder holding it — the part the user chose, with the
/// ancestry every row carries alike removed.
enum RepoSearch {
    /// How the query matched, strongest first. Results sort on this before
    /// anything else: Return opens the first row, so the best match has to be
    /// there no matter which repository is open or most recently used.
    enum Match: Comparable {
        case exactName
        case namePrefix
        case nameSubstring
        case nameInitials
        case nameSubsequence
        case pathSubstring
    }

    /// The strongest way `query` matches the repository at `path`, or `nil`
    /// when it doesn't match at all. Case folding and trimming happen here, so
    /// callers pass the field's text and the store's folders unmodified.
    static func match(query: String, for path: String, scanFolders: [String]) -> Match? {
        let needle = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !needle.isEmpty else { return nil }

        let name = RepoDirectoryStore.displayName(of: path).lowercased()
        if name == needle { return .exactName }
        if name.hasPrefix(needle) { return .namePrefix }
        if name.contains(needle) { return .nameSubstring }
        if initials(of: name).hasPrefix(needle) { return .nameInitials }
        if isSubsequence(needle, of: name) { return .nameSubsequence }
        if searchablePath(for: path, scanFolders: scanFolders).contains(needle) {
            return .pathSubstring
        }
        return nil
    }

    /// First letter of each word: repository names are hyphenated or
    /// underscored, so `gpm` finds `git-projects-manager` without relying on
    /// the far weaker whole-name subsequence.
    private static func initials(of name: String) -> String {
        String(name.split { !$0.isLetter && !$0.isNumber }.compactMap(\.first))
    }

    /// The part of `path` worth searching: whatever lies below the deepest
    /// scan folder (or the home directory) containing it. Everything above
    /// that is common to every row and carries no signal — including it is
    /// what made the old rule match the whole list.
    private static func searchablePath(for path: String, scanFolders: [String]) -> String {
        let roots = (scanFolders + [NSHomeDirectory()])
            .map { $0.hasSuffix("/") ? $0 : $0 + "/" }
            .filter { path.hasPrefix($0) }
        guard let root = roots.max(by: { $0.count < $1.count }) else {
            return path.lowercased()
        }
        return String(path.dropFirst(root.count)).lowercased()
    }

    /// Every character of `needle` appears in `haystack`, in order, not
    /// necessarily adjacent. Both are expected already lowercased.
    private static func isSubsequence(_ needle: String, of haystack: String) -> Bool {
        var index = needle.startIndex
        for character in haystack {
            guard index < needle.endIndex else { break }
            if character == needle[index] {
                index = needle.index(after: index)
            }
        }
        return index == needle.endIndex
    }
}

File: apps/tauri-app/src/lib/utils/repoSearch.ts
/**
 * Ranked fuzzy search over the repository list, shared by the startup picker
 * and the header dropdown — and identical, tier for tier, to the native
 * client's `RepoSearch.swift`, so both frontends answer a query the same way.
 *
 * The rule this replaces accepted the query as a subsequence of the *full
 * path*, which collapses once every row shares an ancestry: under
 * `/Users/leo/Dev/LeoManrique/Desktop`, `llm` matched every repository — the
 * `l` of `leo`, the `l` of `LeoManrique`, and that word's `m` — so the filter
 * appeared not to filter at all.
 *
 * The two halves of a path carry different amounts of signal, so they are
 * searched by different rules: a name keeps the scattered-subsequence match a
 * fuzzy finder is expected to have, while the path must contain the query
 * *contiguously* and is first trimmed to what lies below the scan folder that
 * found it — the part the user chose, with the ancestry every row carries
 * alike removed.
 */

/** How a query matched, strongest first — callers sort ascending. */
export const RepoMatch = {
  ExactName: 0,
  NamePrefix: 1,
  NameSubstring: 2,
  NameInitials: 3,
  NameSubsequence: 4,
  PathSubstring: 5,
} as const

export type RepoMatch = (typeof RepoMatch)[keyof typeof RepoMatch]

/**
 * The strongest way `query` matches a repository, or `null` when it doesn't.
 *
 * @param names Labels to match loosely, best first — a repo's GitHub name and
 *   its `owner/name` where known, otherwise just the folder's basename.
 * @param scanFolders Discovery's `~`-expanded folders, whose prefixes are
 *   stripped from the path before it is searched.
 */
export function matchRepo(
  query: string,
  path: string,
  names: string[],
  scanFolders: string[]
): RepoMatch | null {
  const needle = query.trim().toLowerCase()
  if (!needle) return null

  let best: RepoMatch | null = null
  for (const name of names) {
    const tier = matchName(needle, name.toLowerCase())
    if (tier !== null && (best === null || tier < best)) best = tier
  }
  if (best !== null) return best

  return searchablePath(path, scanFolders).includes(needle) ? RepoMatch.PathSubstring : null
}

function matchName(needle: string, name: string): RepoMatch | null {
  if (name === needle) return RepoMatch.ExactName
  if (name.startsWith(needle)) return RepoMatch.NamePrefix
  if (name.includes(needle)) return RepoMatch.NameSubstring
  if (initials(name).startsWith(needle)) return RepoMatch.NameInitials
  if (isSubsequence(needle, name)) return RepoMatch.NameSubsequence
  return null
}

/**
 * First letter of each word: repository names are hyphenated or underscored,
 * so `gpm` finds `git-projects-manager` without relying on the far weaker
 * whole-name subsequence.
 */
function initials(name: string): string {
  return name
    .split(/[^\p{L}\p{N}]+/u)
    .filter(Boolean)
    .map((word) => word[0])
    .join('')
}

/**
 * The part of `path` worth searching: whatever lies below the deepest scan
 * folder containing it. Everything above that is common to every row and
 * carries no signal — including it is what made the old rule match the whole
 * list. Both separators fold to `/` for the comparison, since discovery
 * returns `C:\Users\…` on Windows.
 */
function searchablePath(path: string, scanFolders: string[]): string {
  const target = normalize(path)
  let cut = 0
  for (const folder of scanFolders) {
    const prefix = normalize(folder).replace(/\/?$/, '/')
    if (prefix.length > cut && target.startsWith(prefix)) cut = prefix.length
  }
  return target.slice(cut)
}

function normalize(path: string): string {
  return path.replace(/\\/g, '/').toLowerCase()
}

/**
 * Every character of `needle` appears in `haystack`, in order, not
 * necessarily adjacent. Both are expected already lowercased.
 */
function isSubsequence(needle: string, haystack: string): boolean {
  let i = 0
  for (let j = 0; j < haystack.length && i < needle.length; j++) {
    if (haystack[j] === needle[i]) i++
  }
  return i === needle.length
}
